README
Copyright (C) 2009 Dennis Schwertel <s@digitalkultur.net>

persy -- sync tool based on Git
=================================
Since persy builds upon the infrastructure offered by Git, it shares its main
strengths:
* speed: recovering your data is faster that cp -a...
* full revision history
* space-efficient data store, with file compression and textual/binary deltas
* efficient transport protocol to replicate the backup (faster than rsync)

Dependencies
============
persy needs the following software at run-time:
* git-core
* python-pyinotify - to get efficient information about filesystem changes
* python-paramiko - ssh library to initialize a remote server

Installation
============
persy can be used just to backup your data or sync your data with multiple 
machines. persy is designed to run by its own on a computer as a revision based
backup application or in an environment with multiple computers and at least 
one server to keep files and folder in sync between them. 


Usage
=====
The normal local workflow is:
 $ persy --config --uname=USERNAME      # username used for the commits
 $ persy --config --mail=MAIL           # useremail used for the commits
 $ persy --init                         # run once to initialize the backup system
 $ persy --config --add_dir=DIR         # DIR is the absolute path to a directory 
                                        # in /home/user (example: /home/username/documents).
                                        # The directory is then integrated in persy
 $ persy --start                        # starts the persy daemon


The normal workflow with a remote Server and an already initialized local repository is:
 $ persy --config --hostname=SERVER     # SERVER = location of the server
 $ persy --config --path=PATH           # PATH = absolute path of the gitrepository on 
                                        # the server (path will be created if it does not exist)
 $ persy --initRemote                   # created a bare git repository on the server in PATH
 $ persy --start                        # starts the persy daemon


The normal workflow with a already initialized remote Server and no local repository. 
IMPORTANT: the synced directories should be empty before the sync. i had some problems 
with already existing files. you can start a sync and then add new files to the synced directory.
 $ persy --config --uname=USERNAME      # username used for the commits
 $ persy --config --email=MAIL          # useremail used for the commits
 $ persy --config --hostname=SERVER     # SERVER = identifier(ip) of the server
 $ persy --config --path=PATH           # PATH = absolute path of the gitrepository on the 
                                        # server
 $ persy --config --add_dir=DIR         # add the same DIR to persy as on the other machines
 $ persy --syncwithremote               # connects to the remote server and 
                                        # pulls the files from the git repository
 $ persy --start                        # starts the persy daemon

License
=======
persy is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 2 of the License, or (at your option) any
later version.

persy is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with persy; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


src/CHANGELOG
persy (0.11.1) lucid; urgency=low

  * fixed show stopper bug with the icons sets on non-ubuntu distributions
  * uses the hostname as the default git commit message

 -- Dennis Schwertel <s@digitalkultur.net>  Sat, 28 Aug 2010 16:00:08 +0200

persy (0.11) lucid; urgency=low

  * added application indicator (ubuntu)
  * old statusicon implemented as fallback
  * redesigned menu

 -- Dennis Schwertel <s@digitalkultur.net>  Sat, 22 Aug 2010 9:00:08 +0200

persy (0.10) lucid; urgency=low

  * small documentation updates
  * added experimental support for remote svn repositories with git-svn

 -- Dennis Schwertel <s@digitalkultur.net>  Sat, 3 Jul 2010 10:30:08 +0200

persy (0.9.1) lucid; urgency=low

  * documentation updates
  * added support for GIT_WORK_TREE and GIT_DIR in the configuration
  * added support for multiple configurations
  * experimental support for fancy remote git servers (like gitoris)
  * bug fixes

 -- Dennis Schwertel <s@digitalkultur.net>  Sun, 14 May 2010 16:57:08 +0200

persy (0.9ubuntu1) lucid; urgency=low

  * Switch to dpkg-source 3.0 (quilt) format

 -- Dennis Schwertel <s@digitalkultur.net>  Sun, 02 May 2010 13:57:08 +0200

persy (0.9) karmic; urgency=low

  * new and improved documentation
  * new build and installation system
  * new logos for linux mint and fedora
  * new ability to easy start persy at login
  * new and easy autorun at login options
  * added support for different remote server ports 
  * added list of submodules (files in submodules are not tracked in persy)
  * possibility to install persy on every linux distribution (make install!)
  * language updates
  * settings gui updates
  * some bugfixes and cleanups

 -- Dennis Schwertel <s@digitalkultur.net>  18 Feb 2010 13:21:00 +0100


persy (0.8.1) karmic; urgency=low

  * added headless mode
  * added username for serverauth
  * some bugfixes and cleanups

 -- Dennis Schwertel <s@digitalkultur.net>  05 Feb 2010 10:21:00 +0100


persy (0.8) karmic; urgency=low

  * rewrote the settings gui
  * added environment tests

 -- Dennis Schwertel <s@digitalkultur.net>  27 Jan 2010 22:10:00 +0100

persy (0.7.2.1) karmic; urgency=low

  * moved the doc building code out of debian/rules and in its own script
  * fingers crossed it will NOW build on launchpad!

 -- Dennis Schwertel <s@digitalkultur.net>  19 Jan 2010 23:10:00 +0100

persy (0.7.2) karmic; urgency=low

  * fixed wrong glade file
  * using pandoc for most of the documentation
  * added help functionality and help files
 
 -- Dennis Schwertel <s@digitalkultur.net>  18 Jan 2010 23:10:00 +0100

persy (0.7.1) karmic; urgency=low

  * fixing missing glade dependency
  * added autoshare ability to the configuration file
  * some bugfixes
 
 -- Dennis Schwertel <s@digitalkultur.net>  16 Jan 2010 12:10:00 +0100

persy (0.7) karmic; urgency=low

  * new settings gui
  * removed initLocal command
  * readme updated
  * added manpage autoupdate
  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  10 Jan 2010 18:10:00 +0100

persy (0.6.2) karmic; urgency=low

  * added a last remote sync when closing persy
  * updated package
  * readme updated

 -- Dennis Schwertel <s@digitalkultur.net>  21 Dec 2009 14:10:00 +0100

persy (0.6.1.3) karmic; urgency=low

  * fixed a problem with a non-existant configuration file
  * added an experimental gui frame (will add more in the future)

 -- Dennis Schwertel <s@digitalkultur.net>  03 Dec 2009 22:28:00 +0100


persy (0.6.1.2) karmic; urgency=low

  * fixed a show-stopper bug, thanks again emorrp1

 -- Dennis Schwertel <s@digitalkultur.net>  26 Nov 2009 22:34:00 +0100


persy (0.6.1.1) karmic; urgency=low

  * fixed a show-stopper bug, thanks emorrp1

 -- Dennis Schwertel <s@digitalkultur.net>  26 Nov 2009 22:20:00 +0100

persy (0.6.1) karmic; urgency=low

  * removed an warning when changing files in a git submodule

 -- Dennis Schwertel <s@digitalkultur.net>  23 Nov 2009 18:20:00 +0100

persy (0.6) karmic; urgency=low

  * The "almost there" release
  * reorganized every single line of code
  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  23 Nov 2009 9:20:00 +0100

persy (0.5) jaunty; urgency=low

  * The "media whore" release
  * localization started
  * readme improved
  * updated configfile with more default values 
  * updated configfile with some explanations
  * replaced a deprecated python-apt function
  * set icon svg size to 128x128px (and fixing a gtk warning with this)
  * implemented delayed error reporting (finaly!)

 -- Dennis Schwertel <s@digitalkultur.net>  27 Oct 2009 10:20:00 +0100

persy (0.4.3) jaunty; urgency=low

  * added a config option for a prefered gui browser
  * added active remote sync
  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  18 Oct 2009 14:20:00 +0100

persy (0.4.2) jaunty; urgency=low

  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  1 Sep 2009 14:20:00 +0100

persy (0.4.1) jaunty; urgency=low

  * added support for short forms of added directories

 -- Dennis Schwertel <s@digitalkultur.net>  25 Aug 2009 20:20:00 +0100

persy (0.4) jaunty; urgency=low

  * fortune support
  * new icons for unsynced and untracked
  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  13 Aug 2009 17:41:00 +0100

persy (0.3.6) jaunty; urgency=low

  * added svg icons for all the stati 
  * added secret cli option to set environment variables in the parent shell
  * improved statusicon tooltip
  * improved display of logs
  * added support for qgit (gitk is still prefered)
  * lots of bugfixes

 -- Dennis Schwertel <s@digitalkultur.net>  11 Aug 2009 12:34:00 +0100



persy (0.3.5.1) jaunty; urgency=low

  * added svg icon 

 -- Dennis Schwertel <s@digitalkultur.net>  06 Aug 2009 23:33:00 +0100


persy (0.3.5) jaunty; urgency=low
 
  * fixed path to VERSION file
  * make the code more robust
  * added GPL in the about dialog
  * added optimize function
  * added new icons for untracked and unsynced files

usr/lib/persy/persy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#License
#=======
#persy is free software: you can redistribute it and/or modify it
#under the terms of the GNU General Public License as published by the Free
#Software Foundation, either version 2 of the License, or (at your option) any
#later version.

#persy is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with persy; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

try:
	import gettext
	#localizations
	LOCALEDIR = '/usr/lib/persy/locale'
	#init the localisation
	gettext.install("messages", LOCALEDIR)
except Exception as e:
	print "I have problems initializing the translations (gettext). \
Will use plain english instead"
	print str(e)

	#check if the _ function is initialized, if not, do a fallback!
	if not _:
		def _(msg):
			"""fallback-function if the original function 
			did not initialize propperly"""
			return msg


try:
	import sys
	from persy_config import PersyConfig
	from persy_core import Core
	from persy_gtk import PersyGtk
	from persy_helper import PersyHelper
	from persy_talker import Talker
	import os
except ImportError as e:
	print _("You do not have all the dependencies:")
	print str(e)
	sys.exit(1)
except Exception as e:
	print _("An error occured when initialising one of the dependencies!")
	print str(e)
	sys.exit(1)

__author__ = "Dennis Schwertel"
__copyright__ = "Copyright (C) 2009, 2010 Dennis Schwertel"




def main(argv):
	"""
	the main function. the command line arguments are processed here and the next actions are started here too
	"""
	args = argv[1:]

	#cli options
	from optparse import OptionParser
	parser = OptionParser(usage = _("use --start to start the daemon"))
	parser.add_option("--start", action="store_true", default=False, help=_("starts persy"))
	parser.add_option("--initremote", action="store_true", default=False, help=_("initializes the remote repository"))
	parser.add_option("--syncwithremote", action="store_true", default=False, help=_("syncs with a remote repository"))
	parser.add_option("--browse", action="store_true", default=False, help=_("start a git browser"))
	parser.add_option("--log", action="store_true", default=False, help=_("prints git log"))
	parser.add_option("--status", action="store_true", default=False, help=_("prints git status"))
	parser.add_option("--ignore", action="store_true", default=False, help=_("recreates list of all ignored files"))
	parser.add_option("--verbose", action="store_true", default=False, help=_("print git output to stdout and set loglevel to DEBUG"))
	parser.add_option("--actions", action="store_true", default=False, help=_("computer-readable actions in persy"))
	parser.add_option("--optimize", action="store_true", default=False, help=_("optimizes the stored files. saves space and improves performance"))
	parser.add_option("--configfile", dest="configfile", default=None, help=_("use a non-default config file"))
	parser.add_option("--config", action="store_true", default=False, help=_("needed flag to change configurations"))
	parser.add_option("--version", action="store_true", default=False, help=_("prints the version"))
	parser.add_option("--uname", dest="uname", default="", help=_("username used in commit"))
	parser.add_option("--mail", dest="mail", default="", help=_("useremail used in commit"))
	parser.add_option("--headless", action="store_true", default=False, help=_("run in headless mode"))
	parser.add_option("--path", dest="path", default="", help=_("path on the server"))
	parser.add_option("--hostname", dest="hostname", default="", help=_("hostname of the remote server"))
	parser.add_option("--add_dir", dest="add_dir", default="", help=_("add local wachted folders"))
	(options, args) = parser.parse_args(args)

	#creat the configuration
	config =  PersyConfig(options.configfile)

	#change in the userhome for all actions
	os.chdir(config.getAttribute('USERHOME'))

	#create programdirectory and a default config file
	if not os.path.exists(config.getAttribute('PERSY_DIR')):
		os.makedirs(config.getAttribute('PERSY_DIR'))

	log = Talker(config, options.verbose) #verbose = print ALL output to stdout
	log.setLevel(options.verbose) #true = debug, false = info set verbosity to show all messages of severity >= DEBUG

	if options.config:
		changed = False
		if options.hostname:
			changed = True
			config['remote']['hostname'] = options.hostname
		if options.path:
			changed = True
			config['remote']['path'] = options.path
		if options.uname:
			changed = True
			config['general']['name'] = options.uname
		if options.mail:
			changed = True
			config['general']['mail'] = options.mail
		if options.add_dir:
			changed = True
			if type(config['local']['watched']) is str:
				if config['local']['watched']:
					config['local']['watched'] = [config['local']['watched'], options.add_dir]
				else:
					config['local']['watched'] = [options.add_dir]
			else:
				config['local']['watched'].append(options.add_dir)
		if changed:
			config.write()
			log.info("writing new config")
		else:
			log.warn(_("nothing changed, maybe wrong attribute names?"))
		sys.exit(0)
	elif options.syncwithremote or options.initremote:
		if options.hostname:
			config['remote']['hostname'] = options.hostname
		if options.path:
			config['remote']['path'] = options.path
		config['remote']['use_remote'] = True
		config.write()


	core = Core()
	core.init(config, log)


	#check if a local repository is initialized:
	if not core.isLocalInitialized():
		core.init_local()

	if options.initremote:
		core.initRemote()
	elif options.version:
		persyhelper = PersyHelper()
		print persyhelper.getSoftwareVersion('persy')
	elif options.syncwithremote:
		core.syncWithRemote()
	elif options.browse:
		core.browse()
	elif options.log:
		core.vcslog()
	elif options.status:
		core.vcsstatus()
	elif options.ignore:
		core.vcsignore()
	elif options.optimize:
		core.optimize()
	elif options.actions:
		#this i used for cli completion
		for opt in parser.option_list:
			print opt.get_opt_string(),
		sys.exit(0)
	else:
		#check if all the icons are present, just warn is something is missing
		filestocheck = (config.getAttribute('ICON_IDLE'), config.getAttribute('ICON_OK'),
			config.getAttribute('ICON_UNSYNCED'), config.getAttribute('ICON_UNTRACKED'),
			config.getAttribute('ICON_WARN'), config.getAttribute('ICON_ERROR'), 
			config.getAttribute('LOGO'))
		fileresults = map(os.path.exists, filestocheck)
		i = 0 
		for fileresult in fileresults:
			if not fileresult:
				log.warn(_("%s file is missing!") % filestocheck[i])
			i += 1


		#START!
		if options.headless:
			core.persy_start()
		else:
			pgtk = PersyGtk()
			pgtk.init(core, config, log, options.start)


if __name__ == '__main__':
	try:
		main(sys.argv)
	except Exception as e:
		print 'Unexpected error: ' + str(e)
		raise
